src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
         /// Try to parse an int out of the reader (as UTF-8).
         /// </summary>
         /// <param name="value">The parsed int value or default if failed.</param>
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out int, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
         /// <returns>True if successfully parsed.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value)
+        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value, char standardFormat = '\0')
         {
             ReadOnlySpan<byte> unread = reader.UnreadSpan;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
+            if (!Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat))
+            {
+                return false;
+            }
+
+            if (consumed < unread.Length)
             {
+                // The parser found a digit it wouldn't consume so we have all valid data.
                 reader.Advance(consumed);
                 return true;
             }
 
-            return TryParseSlow(ref reader, out value);
+            // If we ate all of our unread there may be more valid digits
+            return TryParseSlow(ref reader, out value, standardFormat);
         }
 
-        private static unsafe bool TryParseSlow(ref BufferReader<byte> reader, out int value)
+        private static unsafe bool TryParseSlow(ref BufferReader<byte> reader, out int value, char standardFormat)
         {
-            const int MaxLength = 15;
-
-            ReadOnlySpan<byte> unread = reader.UnreadSpan;
+            // Note that ints can have leading zeros ("-00000000000000127") which makes predicting the max
+            // length difficult. Typically the int will be less than 15 characters ("-2,147,483,648") so we'll
+            // work in those increments.
+            int bufferSize = 15;
 
-            if (unread.Length > MaxLength)
+            // Try using the stack for scratch space.
+            do
             {
-                // Fast path had enough space but couldn't find valid data
-                value = default;
-                return false;
-            }
-
-            byte* buffer = stackalloc byte[MaxLength];
-            Span<byte> tempSpan = new Span<byte>(buffer, MaxLength);
-            if (Utf8Parser.TryParse(reader.PeekSlow(tempSpan), out value, out int consumed))
+                // stackalloc is separated as we can't represent that PeekSlow will not capture the span.
```

Comment:
represent -> guarantee?

Reply:
represent- we can technically guarantee, we just have no way to represent that to C# (yet)